File: matmul/matmul.lua
local function matgen(n)
    local tmp = 1.0 / n / n;
    local m = {};
    for i = 1, n do
        local a = {};
        for j = 1, n do
            a[j] = tmp * (i - j) * (i + j - 2);
        end
        m[i] = a;
    end
    return m;
end

local function matmul(a, b, nn)
    local m = nn; -- a:len();
    local n = nn; -- a[1]:len();
    local p = nn; -- b[1]:len();

    local b2 = {};
    for i = 1, n do
        b2[i] = {};
        for j = 1, p do
            b2[i][j] = 0;
        end
    end

    for i = 1, n do
        for j = 1, p do
            b2[j][i] = b[i][j];
        end
    end

    local c = {};
    for i = 1, m do
        ci = {};
        ai = a[i];

        for j = 1, n do
            local s = 0;
            local b2j = b2[j];

            for k = 1, p do
                s = s + ai[k] * b2j[k];
            end

            ci[j] = s;
        end

        c[i] = ci;
    end

    return c;
end

local function run(N)
    local start_time = os.clock();
    local a = matgen(N);
    local b = matgen(N);
    local c = matmul(a, b, N);

    local v = c[N / 2 + 1][N / 2 + 1];
    io.write(string.format("%0.9f\n",v));
    io.stderr:write(string.format("time(%.9f)\n", os.clock() - start_time));    
end

local N = tonumber(arg and arg[1]) or 10
local times = tonumber(arg and arg[2]) or 1

io.stderr:write("started")

for i=1,times,1 do
  run(N)
end

File: matmul/matmul.js
var matrix = {}

matrix.new = function (n) {
    var a = new Array(n);
    for (var i = 0; i < n; i++) a[i] = new Float64Array(n);
    return a;
}

matrix.T = function (a, n) {
    var y = matrix.new(n)
    for (var i = 0; i < n; i++) {
        for (var j = 0; j < n; j++) {
            y[i][j] = a[j][i]
        }
    }
    return y
}

matrix.mul = function (a, b, n) {
    var y = matrix.new(n)
    var c = matrix.T(b, n)
    for (var i = 0; i < n; i++) {
        for (var j = 0; j < n; j++) {
            var sum = 0;
            for (var k = 0; k < n; k++) sum = sum + a[i][k] * c[j][k]
            y[i][j] = sum;
        }
    }
    return y
}

matgen = function (n) {
    var y = matrix.new(n)
    var tmp = 1 / n / n
    for (var i = 0; i < n; i++) {
        for (var j = 0; j < n; j++) {
            y[i][j] = tmp * (i - j) * (i + j)
        }
    }
    return y
}

function main(n) {
    start = new Date();
    var a = matgen(n);
    var b = matgen(n);
    var c = matrix.mul(a, b, n);
    var v = c[(n/2)][(n / 2)];
    console.log(v.toFixed(9));
    console.error("time(%d)", ((new Date()) - start) / 1000);
}

(async function() {
    const n = parseInt(process.argv[2]) || 10;
    const times = parseInt(process.argv[3]) || 1;

    console.error("started");
    for (let i = 0; i < times; i++) { main(n); }
})();
